Common/Hashmaps.h

```diff
@@ -0,0 +1,271 @@
+		// No? Let's go into search mode. Linear probing.
+		uint32_t p = pos;
+		while (true) {
+			if (map[p].state == BucketState::TAKEN && KeyEquals(key, map[p].key))
+				return map[p].value;
+			else if (map[p].state == BucketState::FREE)
+				return nullptr;
+			p = (p + 1) & mask;  // If the state is REMOVED, we just keep on walking. 
+			if (p == pos)
+				DebugBreak();
+		}
+		return nullptr;
+	}
+
+	// Returns false if we already had the key! Which is a bit different.
+	bool Insert(const Key &key, Value value) {
+		// Check load factor, resize if necessary. We never shrink.
+		if (count_ > capacity_ / 2) {
+			Grow();
+		}
+		uint32_t mask = capacity_ - 1;
+		uint32_t pos = HashKey(key) & mask;
+		uint32_t p = pos;
+		while (true) {
+			if (map[p].state == BucketState::TAKEN) {
+				if (KeyEquals(key, map[p].key)) {
+					DebugBreak();  // Bad! We already got this one. Let's avoid this case.
+					return false;
+				}
+				// continue looking....
+			} else {
+				// Got a place, either removed or FREE.
+				break;
+			}
+			p = (p + 1) & mask;
+			if (p == pos) {
+				// FULL! Error. Should not happen thanks to Grow().
+				DebugBreak();
+			}
+		}
+		map[p].state = BucketState::TAKEN;
+		map[p].key = key;
+		map[p].value = value;
+		count_++;
+		return true;
+	}
+
+	void Remove(const Key &key) {
+		uint32_t mask = capacity_ - 1;
+		uint32_t pos = HashKey(key) & mask;
+		uint32_t p = pos;
+		while (map[p].state != BucketState::FREE) {
+			if (map[p].state == BucketState::TAKEN && KeyEquals(key, map[p].key)) {
+				// Got it! Mark it as removed.
+				map[p].state = BucketState::REMOVED;
+				count_--;
+				return;
+			}
+			p = (p + 1) & mask;
+			if (p == pos) {
+				// FULL! Error. Should not happen.
+				DebugBreak();
+			}
+		}
+	}
+
+	size_t size() const {
+		return count_;
+	}
+
+	// TODO: Find a way to avoid std::function. I tried using a templated argument
+	// but couldn't get it to pass the compiler.
+	inline void Iterate(std::function<void(const typename Key &key, typename Value value)> func) {
+		for (auto &iter : map) {
+			if (iter.state == BucketState::TAKEN) {
+				func(iter.key, iter.value);
+			}
+		}
+	}
+
+	void Clear() {
+		// TODO: Speedup?
+		map.clear();
+		map.resize(capacity_);
+	}
+
+private:
+	void Grow() {
+		// We simply move out the existing data, then we re-insert the old.
+		// This is extremely non-atomic and will need synchronization.
+		std::vector<Pair> old = std::move(map);
+		capacity_ *= 2;
+		map.clear();
+		map.resize(capacity_);
+		count_ = 0;  // Insert will update it.
+		for (auto &iter : old) {
+			if (iter.state == BucketState::TAKEN) {
+				Insert(iter.key, iter.value);
+			}
+		}
+	}
+	struct Pair {
+		BucketState state;
+		Key key;
+		Value value;
+	};
+	std::vector<Pair> map;
+	int capacity_;
+	int count_ = 0;
+};
+
+// Like the above, uses linear probing for cache-friendliness.
+// Does not perform hashing at all so expects well-distributed keys.
+template <class Value>
+class PrehashMap {
```

Comment:
Hmm, I feel like we could have an abstract parent with 80% of the code, and have each of these be a thin subclass with some template params (i.e. KeyStorage template param or something.)

-[Unknown]